ext/opcache/Optimizer/block_pass.c

```diff
@@ -1335,185 +1351,153 @@ static void zend_jmp_optimization(zend_basic_block *block, zend_op_array *op_arr
 
 		case ZEND_JMPNZ_EX:
 		case ZEND_JMPZ_EX:
-			if (block->successors[0] == block->successors[1]) {
-				/* L: T = JMP[N]Z_EX(X, L+1) -> T = BOOL(X) */
-
-				last_op->opcode = ZEND_BOOL;
-				last_op->op2.num = 0;
-				block->successors_count = 1;
-				++(*opt_count);
-				break;
-			}
+optimize_jmpznz:
+			jmp_hitlist_count = 0;
 
-			if (1) {
-				zend_op *target, *target_end;
-				zend_basic_block *target_block;
-				zend_uchar same_type = last_op->op1_type;
-				uint32_t same_var = last_op->op1.var;
-				uint32_t same_res_var = last_op->result.var;
-
-				target_block = blocks + block->successors[0];
-next_target_ex:
+			target_block = get_target_block(cfg, block, 0, opt_count);
+			while (target_block->len == 1) {
 				target = op_array->opcodes + target_block->start;
-				target_end = target + target_block->len;
-				while (target < target_end && target->opcode == ZEND_NOP) {
-					target++;
-				}
- 				/* next block is only NOP's */
-				if (target == target_end) {
-					target_block = blocks + target_block->successors[0];
-					++(*opt_count);
-					goto next_target_ex;
+
+				if (target->opcode == ZEND_JMP) {
+					/* T = JMPZ_EX(X, L), L: JMP(L2) -> T = JMPZ(X, L2) */
+					next = target_block->successors[0];
 				} else if (target->opcode == last_op->opcode-3 &&
-				           ((same_var == target->op1.var &&
-				             same_type == target->op1_type) ||
-				            (same_res_var  == target->op1.var &&
-				             target->op1_type == IS_TMP_VAR)) &&
-				           !(target_block->flags & ZEND_BB_PROTECTED)) {
+				           (SAME_VAR(target->op1, last_op->result) ||
+				            SAME_VAR(target->op1, last_op->op1))) {
 					/* T = JMPZ_EX(X, L1), L1: JMPZ({X|T}, L2) -> T = JMPZ_EX(X, L2) */
```

Comment:
I think the worst thing that can happen is an error handler actually changing the value, in which case the control flow will be incorrect:

```
<?php
  
set_error_handler(function() {
    $GLOBALS['x'] = true;
});

if ($x) echo "foo\n";
if ($x) echo "bar\n";
```

This will print "bar" without opcache and nothing with. But this also requires "malicious" code.

Theoretically cast_object handlers could also be an issue, but I don't think we have any implementations that could return two different values on directly subsequent calls.

Reply:
Anyway, this bug is independent from the PR and exists for ages.
We may decide to fix it in 7.3, 7.4, in master or not to fix at all.
I prefer to land this without fixing it.